magda-indexer/src/main/scala/au/csiro/data61/magda/indexer/helpers/StreamSourceController.scala

```diff
@@ -0,0 +1,129 @@
+  */
+class StreamSourceController(bufferSize: Int, streamController: StreamController)
+                            (implicit val system: ActorSystem,
+                             implicit val config: Config,
+                             implicit val materializer: Materializer) {
+
+  val log = Logging(system, getClass)
+  implicit val ec: ExecutionContextExecutor = system.dispatcher
+  implicit val scheduler: Scheduler = system.scheduler
+
+  private val GET_MORE_DATASETS: String = "Get more datasets"
+  private val NO_MORE_DATASETS: String = "No more datasets"
+  private val dataSetCount = new AtomicLong(0)
+  private val (ref, source) = createStreamSource()
+
+  /**
+    * A stage that acts upon control messages (either getting more datasets or completing the stream).
+    *
+    * @tparam Object a generic type for either DataSet or String types.
+    */
+  private final case class MessageChecker[Object]() extends SimpleLinearGraphStage[Object] {
+    override def initialAttributes: Attributes = DefaultAttributes.take
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        override def onPush(): Unit = {
+          val message = grab(in)
+          if (message == NO_MORE_DATASETS) {
+            completeStage()
+            ref ! PoisonPill
+          }
+          else if (message == GET_MORE_DATASETS) {
+            if (streamController != None.orNull)
+              streamController.next(bufferSize / 2)
+
+            pull(in)
+          }
+          else
+            push(out, message)
+        }
+
+        override def onPull(): Unit = {
+          pull(in)
+        }
+
+       setHandlers(in, out, this)
+      }
+
+    override def toString: String = "MessageChecker"
+  }
+
+  private def createStreamSource(): (ActorRef, Source[DataSet, NotUsed]) = {
+    // The additional 2 slots in the buffer are for the the control messages GET_MORE_DATASETS
+    // and NO_MORE_DATASETS.
+    val (ref: ActorRef, publisher: Publisher[DataSet]) =
+      Source.actorRef[DataSet](bufferSize = bufferSize + 2, OverflowStrategy.fail)
+        .toMat(Sink.asPublisher(true))(Keep.both).run()
+
+    val source: Source[DataSet, NotUsed] = Source.fromPublisher(publisher)
+    val source2: Source[DataSet, NotUsed] = source.via(MessageChecker.apply())
+    (ref, source2)
+  }
+
+  /**
+    * Fill the source with given datasets
+    *
+    * @param dataSets a batch of datasets to fill the source.
+    * @param hasNext indicating if there are more datasets available
+    * @param isFirst indicating if this is the first batch of datasets
+    */
+  def fillSource(dataSets: Seq[DataSet], hasNext: Boolean, isFirst: Boolean): Unit = {
+    if (isFirst){
+      dataSets foreach (dataSet => {
+        dataSet.copy(publisher = dataSet.publisher)
+        ref ! dataSet
+        if (dataSetCount.incrementAndGet() == bufferSize / 2 && hasNext)
```

Comment:
The logic (that determines the timing of sending `ref ! GET_MORE_DATASETS` message) probably should be moved outside the foreach loop. The condition probably should not be `== bufferSize / 2` --- currently, if indexer fetches less than `bufferSize / 2` records in one request (indexer should fetch much smaller number of dataset in a single request max. 50 as previously comment), the crawling process will hang there.

Reply:
Have you verified it?

Reply:
Yes. You are right --- I missed `dataSetCount.incrementAndGet() % (bufferSize / 2) == 0` part and was wondering where we decrease the `dataSetCount`.
Had a few more tests last night and logic here doesn't seems to create any troubles. 👍 
I've also reduced bufferSize to a lower value (100) and the new logic does stops crawling more while elasticsearch was still processing:
```
[INFO] [12/04/2018 09:55:15.824] [indexer-akka.actor.default-dispatcher-10] [RegistryCrawler(akka://indexer)] Crawling registry at /v0
[INFO] [12/04/2018 09:55:22.074] [indexer-akka.actor.default-dispatcher-10] [StreamController(akka://indexer)] Total crawled 1000 datasets from registry
[INFO] [12/04/2018 09:55:24.301] [indexer-akka.actor.default-dispatcher-16] [StreamController(akka://indexer)] Total crawled 1500 datasets from registry
[INFO] [12/04/2018 09:55:24.487] [indexer-akka.actor.default-dispatcher-14] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 1 datasets
[INFO] [12/04/2018 09:55:55.156] [indexer-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:55:56.711] [indexer-akka.actor.default-dispatcher-2] [StreamController(akka://indexer)] Total crawled 2000 datasets from registry
[INFO] [12/04/2018 09:56:17.545] [indexer-akka.actor.default-dispatcher-17] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:56:18.978] [indexer-akka.actor.default-dispatcher-2] [StreamController(akka://indexer)] Total crawled 2500 datasets from registry
[INFO] [12/04/2018 09:56:51.148] [indexer-akka.actor.default-dispatcher-10] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:56:52.707] [indexer-akka.actor.default-dispatcher-8] [StreamController(akka://indexer)] Total crawled 3000 datasets from registry
[INFO] [12/04/2018 09:57:22.418] [indexer-akka.actor.default-dispatcher-2] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:57:24.139] [indexer-akka.actor.default-dispatcher-15] [StreamController(akka://indexer)] Total crawled 3500 datasets from registry
[INFO] [12/04/2018 09:57:38.986] [indexer-akka.actor.default-dispatcher-32] [akka.actor.ActorSystemImpl(indexer)] Successfully indexed 500 datasets
[INFO] [12/04/2018 09:57:40.657] [indexer-akka.actor.default-dispatcher-15] [StreamController(akka://indexer)] Total crawled 4000 datasets from registry
```
Well done & great job~ 🎉